avalon/pipeline.py

```diff
@@ -1325,31 +1333,36 @@ def path_from_config():
             "representation": representation["name"],
             "user": Session.get("AVALON_USER", getpass.getuser()),
             "app": Session.get("AVALON_APP", ""),
-            "task": Session.get("AVALON_TASK", "")
+            "task": Session.get("AVALON_TASK", ""),
+            "hierarchy": asset["data"].get("hierarchy", "")
         }
 
         try:
             path = template.format(**data)
         except KeyError as e:
-            log.debug("Template references unavailable data: %s" % e)
+            log.warning("Missing value for filepath template: {}".format(e))
             return None
 
         if os.path.exists(path):
             return os.path.normpath(path)
-
-    def path_from_data():
-        if "path" not in representation["data"]:
+        else:
+            log.warning("path from config doesn't exist: {}".format(path))
             return None
 
-        path = representation["data"]["path"]
+    def path_from_data():
+        path = representation["data"].get("path", None)
         if os.path.exists(path):
             return os.path.normpath(path)
+        else:
+            log.warning("path from data doesn't exist: {}".format(path))
+            return None
 
-    return (
-        path_from_represenation() or
-        path_from_config() or
-        path_from_data()
-    )
+    resolved_path = (path_from_representation()
+                     or path_from_config()
+                     or path_from_data())
+
+    assert resolved_path, "Cannot resolve path!"
```

Comment:
Not sure if adding this assertion now is a good idea. One may already checking path resolved by return value is `None` or not. Adding assertion would mean forcing everyone to handle `AssertionError` on top of current usage after this. No ?

One alternative would be:

```python
def get_representation_path(representation, error_out=False):
    ...
    if error_out:
        assert resolved_path, "Cannot resolve path!"
    return resolved_path

```

Reply:
Yes, definitely don't introduce new exceptions. Debug log it instead, that's for developers whereas new exceptions are raised all the way to the artist.

Reply:
I can agree with this being a core API function thus not throwing additional error/exceptions.

I've moved this over to the loader initialization, where a valid path is required for Loader to behave properly. If that happens, it means something went horribly wrong with the asset being loaded. So I think it's reasonable to error out with assert failure.

Reply:
An assert isn't the right approach here. This will still propagate an error to the user, and the user won't be the one to deal with it. At this point, the onus is on Avalon to handle and format any error that occurs in a way that makes sense to the user. If the Loader is to have an assert, than whomever instantiates it needs to handle that assert.

In this case, if a filename really is invalid, the problem is further up the pipeline. Either an asset wasn't published correctly, or a physical path that used to exist exists no more. That's what we can tell the end user.

In practice, it looks like this fname isn't actually used?

https://github.com/getavalon/core/blob/91f5ae87120f61d8beb57af099d321e28178399f/avalon/maya/pipeline.py#L554-L561

So we'd be throwing an assert error for no reason. Let me know what you think.

Reply:
Right, an assert failure is probably too blunt for the end user, though I don't quite see how the current instance of Loader can  gracefully function when there's no valid filename. ( Hmm I guess it's very unlikely to happen, but hopefully this discussion is interesting in theory..) 

A bit more details about what led me here: 
I was hacking and experimenting with the AVA config, where I included {hierarchy} in the filepath template. The very first error I saw was on Line 559 in the code snippet above, saying something like `'None' type has no .replace()`. So I put the line into a ` if self.fname: ` block, to move on when there's no `fname`. That's when I got the Maya error I initially posted in the Gitter chat. This `self.fname` will be passed to cmds as an argument, having no value would give that Maya error popup.

So, consider we create an Loader instance, and the `context`  provided is broken somehow, and causing `self.fname` to be None. Then this loader instance can't fulfill its mission, since it can't figure out which file to load. 

I was also sorting out thoughts while I'm typing this, I now think maybe the initialization code of Loader class should handle possible exception, since the 'context' it provides may result in a broken Loader. 

Sorry for the long-winded reply. If the changes required are too complicated, I don't mind going with whatever you suggest at all.  =)

Reply:
Agree that `'None' type has no .replace()` is confusing for everyone. I think just add an if-statement to check `self.fname is None` is enough.

The `fname` attribute in `Loader.__init__()` is just a start point for developer to implement the rest of the function of loader plugin, e.g.
https://github.com/getavalon/core/blob/91f5ae87120f61d8beb57af099d321e28178399f/avalon/pipeline.py#L178-L190

So, yeah, if you prefer raising AssertionError or any other action, that should be in your own implemented loader plugin method e.g. `Loader.load()`

Reply:
> though I don't quite see how the current instance of Loader can gracefully function when there's no valid filename

I think the proper approach would be to catch this *before* it gets to the loader. Who's passing an invalid filename to it? And why? That's what we've got to catch and handle, at the source. So that's something I think would be welcomed and and encouraged!

The debug messages are still great though, because it allows you the developer to enable debug messages and gain insight into what's going on. It's just important to keep these two audiences in mind, and provide relevant information to the relevant party.

Reply:
> Right, an assert failure is probably too blunt for the end user, though I don't quite see how the current instance of Loader can gracefully function when there's no valid filename. ( Hmm I guess it's very unlikely to happen, but hopefully this discussion is interesting in theory..)

It could very well be that the Loader itself does some custom logic based on the context data that it in the end does get an existing path and that it could disregard `self.fname` completely. So technically, it could already be used.

I agree however that the case you hit should come with more graceful handling. I'd say in your case it sounds like the particular loader could handle it more gracefully with a message to the user maybe?

Reply:
Thank you guys for the input. Now the Maya Loader simply log an warning when the fname is None.

I was thinking of adding an assertion in the `avalon.maya.compat.load()`, which was on my original stacktrace. But just realized this function is deprecated. So I'll let it be.